src/jit/codegenxarch.cpp

```diff
@@ -2470,6 +2470,18 @@ CodeGen::genCodeForTreeNode(GenTreePtr treeNode)
     case GT_JTRUE:
         {
             GenTree *cmp = treeNode->gtOp.gtOp1;
+
+#if !defined(_TARGET_64BIT_)
+            GenTreePtr op1 = cmp->gtGetOp1();
+            bool isSigned = ((cmp->gtFlags & GTF_UNSIGNED) == 0);
+
+            // For long signed compares that are not GT_NE or GT_EQ, we emit the logic in genCompareLong,
+            // and do not need to do anything here.
+            if (varTypeIsLong(op1->TypeGet()) && isSigned && (cmp->gtOper != GT_NE && cmp->gtOper != GT_EQ))
+            {
+                break;
+            }
+#endif
             assert(cmp->OperIsCompare());
```

Comment:
While I'm ok with x86 doing all the work in genCompareLong() for longs for both GT_CMP and GT_JTRUE cases, it still feels slightly wrong to me. It seems like we should have a function that is flexible enough to handle either case, and call it from the JTRUE case or the CMP case. Something to think about.